### PR TITLE
Reduce number of dropout test case

### PR DESCRIPTION
**Pull request by ce1adon**
To save CI resource
Issue #1128

### CODE REVIEW

File: test/dropout.cpp
Comment (junliume) [inline test/dropout.cpp:42]: @ce1adon Global non-const variables are evil :)

Comment (ce1adon) [reply test/dropout.cpp:42]: Hmm I can't figure out another way to count the number here. Any suggestion?

Comment (junliume) [reply test/dropout.cpp]: Hi @atamazov  could you chip in to help please? Actually there are two options:
-  Run this dropout test exactly once. I would recommend moving ```static int num_test_case = 0;``` into scope of ```struct dropout_driver : test_driver```, I think we'll have a single instance of static variable among all translation units.
-  I suspect the following random vector filling has been called many times
   ```
   srand(0);
        for(size_t i = 0; i < in.desc.GetElementSize(); i++)
             reserveSpace[i] =
                 static_cast<unsigned char>(float(GET_RAND()) / float(RAND_MAX) > dropout_rate);
   ```
   I have not done the profiling myself, I would guess would this be a perf bottleneck **especially with multi-threading**? Could we fill it once only, and only shuffle in other instances? Or better yet, use other random generator?

Comment (ce1adon) [reply test/dropout.cpp]: Solved with macro

**Comment by junliume** [comment]
@ce1adon  Thank you!

Fp16 Hip Install All Vega20
> [2021-09-01T18:49:30.258Z]  1/81 Test #17: test_dropout ...............................   Passed    5.03 sec

Fp32 Hip All Vega20
> [2021-09-01T18:45:12.918Z]  1/82 Test #17: test_dropout ...............................   Passed    4.86 sec

CC: @atamazov  @JehandadKhan

**Review by junliume — APPROVED** [review]

File: test/dropout.cpp
Comment (atamazov) [inline test/dropout.cpp:37]: [Recommendation] It should be like this:
```C++
#define WORKADOUND_ISSUE_1128 1
...
#if WORKADOUND_ISSUE_1128
#define DROPOUT_SINGLE_CTEST 1
#endif
...
```

**Comment by atamazov** [comment]
@ce1adon @junliume Does it affect the quality of testing? In other words, it this really a workaround? If yes, then what it the proposed full-blown fix?

**Comment by junliume** [comment]
> @ce1adon @junliume Does it affect the quality of testing? In other words, it this really a workaround? If yes, then what it the proposed full-blown fix?

The rational is that bottleneck is filling input_dims numerous times with random numbers instead of actually test dropout functionality. Aside from other drop out tests (e.g. test_gru_dropout test_lstm_dropout test_rnn_vanilla_dropout), it should be sufficient to check dropout once here.
CC: @ce1adon to double check the above statement.

**Comment by ce1adon** [comment]
@atamazov Hey Artem, maybe we should name it reduce number of tests instead of workaround. Please notice that what we did is to reduce test number to the limit. A real "workaround" to the test would be a complete change of the test rule to test RNG path and dropout path separately. Hope this help your evaluation.

**Comment by atamazov** [comment]
@ce1adon 
> @atamazov Hey Artem, maybe we should name it reduce number of tests instead of workaround. 

This question is: Does this (reducing the number of tests) affects the quality of testing?

> A real "workaround" to the test would be a complete change of the test rule to test RNG path and dropout path separately.

Can you please explain a little bit what is the purpose of the separation of RNG and dropout? In other words, why we need it? Please put this explanation in #1128. Thanks.

**Comment by ce1adon** [comment]
> This question is: Does this (reducing the number of tests) affects the quality of testing?

I would say no but I won't be surprised that different people have different opinions.

> Can you please explain a little bit what is the purpose of the separation of RNG and dropout? In other words, why we need it? Please put this explanation in #1128. Thanks.

We can do that later. I explained it many times and got approval but there are always something with higher priority blocking this change.

**Comment by atamazov** [comment]
@ce1adon 
> I would say no but I won't be surprised that different people have different opinions.

Your opinion is the most important for me. So, this PR is really a fix (not a workaround). Good!

> I explained it many times and got approval...

Okay. It seems that I somehow missed the explanations. I would appreciate if you provide me with a link to the explanation (if it is somewhere on github). If not, then no problem.

**Pull request merged**
